libusb/os/windows_winusb.c

```diff
@@ -2017,6 +2017,157 @@ static int winusb_get_device_list(struct libusb_context *ctx, struct discovered_
 	return r;
 }
 
+struct string_descriptor_s {
+	UCHAR bLength;
+	UCHAR bDescriptorType;
+	uint8_t bString[252];  // UTF-16LE, must be even length
```

Comment:
I think it'd be safer to use `uint16_t bString[126]`.  You're going to want `bString` to be aligned to the size of uint16, not uint8.

Reply:
USB descriptors are usually represented as arrays of bytes.  While the struct will pack correctly with uint16_t, it seems odd to me.  The purpose to to read bytes back through a USB control message.  It uses `usbi_utf16le_to_utf8` to convert from UTF-16LE to UTF-8, and `usbi_utf16le_to_utf8` also takes bytes for the UTF-16LE.

How does changing type change memory safety?  We never access the UTF-16LE as uint16_t.

Reply:
>While the struct will pack correctly with uint16_t...

I'm pretty sure compilers are free to add any padding they want to a struct. You can't just assume the struct will be packed tightly. And if the field is typed as `uint8_t` then it only needs to be aligned on an 8 byte boundary. By using `uint16_t bString[126]` you guarantee that it's aligned to 16 bits. You can then safely cast to `uint8_t*` elsewhere.  The opposite cast is not necessarily valid and is what `-Wcast-align` warns about.

Reply:
Well, according to C17 6.7.2.1 comment 15:

> Each non-bit-field member of a structure or union object is aligned in an implementation-defined
> manner appropriate to its type.

However, in practice, all major compilers align members to their type width and add padding before if needed.  In this structure, there is no padding since all types are `u8`.  We *REQUIRE* this behavior since the bytes coming back from the CONTROL IN request are directly placed into this structure, and we use this structure to "parse" the IN response.  Changing to uint16_t actually wouldn't matter here since the two UCHAR ensure uint16_t alignment.

However, there are no `Wcast-align`  errors since the code manually unpacks these bytes as little endian UTF-16LE code points.  See above for the big difference between `(uint16_t *) src` (which the code does not do) and `(uint16_t) *src` (which the code does).

uint16_t is not really correct either, at least on big-endian machines (not that any still exist that we care about here).

Since we always access this as bytes, I see only more confusion going to uint16_t.

Does this make sense?  Do you agree?

Reply:
>We REQUIRE this behavior since the bytes coming back from the CONTROL IN request are directly placed into this structure

In that case, the struct should be tagged with `LIBUSB_PACKED`.

Yeah, sorry about the cast confusion.

Reply:
Added LIBUSB_PACKED.  However, this addition does not change anything in this case since all members are uint8.

Reply:
>However, this addition does not change anything in this case since all members are uint8.

With today's compiler. But again, the compiler is free to add padding if it wants, so the packing is required for correctness. Yeah, I'm being pedantic. :)

But I've just noticed that `LIBUSB_PACKED` is actually only defined for GCC:

```
#if defined(__GNUC__)
#define LIBUSB_PACKED __attribute__ ((packed))
#else
#define LIBUSB_PACKED
#endif /* __GNUC__ */
```

IIRC MS doesn't support a struct attribute like that.  Since this file is Windows only, I guess just add:

```
#ifdef _MSC_VER
#pragma pack(push, 1)
#endif
```

around your struct.

GCC/Clang does support `#pragma pack` though, so we could create new macros `LIBUSB_PACKED_PUSH` / `LIBUSB_PACKED_POP` that support all 3 compilers... that can be another PR though. :)  I could do that...

Reply:
The use of `LIBUSB_PACKED` sounds like a larger issue.  A bunch of structs in libusbi.h have `LIBUSB_PACKED`.  At least adding `LIBUSB_PACKED` to string_descriptor_s is consistent with everywhere else.  I don't think that we should add the `#pragma pack(push, 1)` here, since it is not strictly needed and makes this different.

If you want a new PR to improve struct packing, go for it after this pull request gets merged!  Any day now?  I know I'm ready to get this done ;)

Reply:
It works great... no pack needed on any compiler when all members are bytes.  In the last several decades, the only use for struct pack is when you want a member to be unaligned.

This pull request was a complete implementation of a twice-abandoned feature, now with support for all standard strings (vendor, product, serial number) that blocked the previous merges.  This does add an API call, but really... *months*.

Thank you for your review and feedback.

Reply:
>This pull request was a complete implementation of a twice-abandoned feature, now with support for all standard strings (vendor, product, serial number) that blocked the previous merges.

Yeah, it looks great!

>This does add an API call, but really... months.

Well, it's open source. Volunteer time is up to the volunteers.

>no pack needed on any compiler when all members are bytes

I don't think that's necessarily true. C99 6.7.2.1: "Each non-bit-field member of a structure or union object is aligned in an implementation-defined manner appropriate to its type."  Better to explicitly mark to packed to be sure.

Reply:
Yes, I added `LIBUSB_PACKED`, which currently does nothing for MSVC.  MSVC does the right thing, and this structure is no different from every other structure in libusbui.h marked with `LIBUSB_PACKED` that needs back-to-back members.  Improving this is your pull request https://github.com/libusb/libusb/pull/1542.

Reply:
The other structures in libusbi.h are indeed marked with `LIBUSB_PACKED` but they are ***also*** wrapped in:

```
#ifdef _MSC_VER
#pragma pack(push, 1)
#endif
```

I would use this same pragma in addition to the `LIBUSB_PACKED` you already added.

Reply:
Done.